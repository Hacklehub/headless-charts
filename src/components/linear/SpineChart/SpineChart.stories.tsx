import { Meta } from '@storybook/react';
import SpineChart from '.';

export default {
  title: 'Linear/SpineChart',
  component: SpineChart,
  tags: ['autodocs'],
} as Meta;

export const Default = {
  args: {
    data: [
      {
        value: 5,
      },
      {
        value: 7,
      },
      {
        value: 9,
      },
    ],
    id: 'spine-chart-default',
    y: {
      key: 'axis-value',
      direction: 'left',
      className: '',
    },
    x: [
      { key: 'x1', direction: 'left', className: 'left-class' },
      { key: 'x2', direction: 'right', className: 'right-class' },
    ],
  },
};
