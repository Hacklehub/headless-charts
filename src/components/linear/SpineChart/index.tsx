import { axisBottom, axisLeft, axisRight, axisTop } from 'd3-axis';
import { defaultChartClassNames, mergeTailwindClasses } from '../../../utils';
import { max, sum } from 'd3-array';
import { scaleBand, scaleLinear } from 'd3-scale';
import { select, selectAll } from 'd3-selection';
/* eslint-disable @typescript-eslint/no-explicit-any */
import { useCallback, useEffect } from 'react';

interface Y {
  key: string;
  direction: 'left' | 'right';
  className: string;
}

interface DataItem {
  [key: string]: number | string;
}

interface SpineChartProps {
  data: DataItem[];
  id: string;
  className?: string;
  paddingBar?: number;
  padding?: {
    left: number;
    right: number;
    bottom: number;
    top: number;
  };
  margin?: {
    left: number;
    right: number;
    top: number;
    bottom: number;
  };
  y: Y;
  x: Array<any>;
}

const SpineChart = ({
  data = [],
  id,
  className,
  paddingBar = 0.3,
  padding = {
    left: 0,
    right: 0,
    top: 0,
    bottom: 0,
  },
  margin = {
    top: 40,
    bottom: 40,
    left: 40,
    right: 20,
  },
  y,
  x,
}: SpineChartProps) => {
  const refreshChart = useCallback(() => {
    const svg = select(`#${id}`);
    svg.selectAll('*').remove();

    const width = +svg.style('width').split('px')[0],
      height = +svg.style('height').split('px')[0];

    const g = svg.append('g');

    const leftSeries = x.filter((column) => column.direction === 'left'),
      rightSeries = x.filter((column) => column.direction !== 'left');

    /* eslint-disable */
    // @ts-ignore
    const extreme = max([
      // @ts-ignore
      max(data.map((row) => sum(leftSeries.map((column) => row[column.key])))),
      // @ts-ignore
      max(data.map((row) => sum(rightSeries.map((column) => row[column.key])))),
    ]);

    const halfWidth =
      (width - padding.left - margin.left - padding.right - margin.right) / 2;

    const xRightFn = scaleLinear()
      // @ts-ignore
      .domain([0, extreme])
      .range([padding.left + margin.left + halfWidth, width - margin.right]);

    const xLeftFn = scaleLinear()
      // @ts-ignore
      .domain([0, extreme])
      .range([padding.left + margin.left + halfWidth, padding.left + margin.left]);

    const xLeftAxis =
      // @ts-ignore
      x.axis === 'top'
        ? // @ts-ignore
          axisTop(xLeftFn).ticks(x.axisTicks)
        : // @ts-ignore
          axisBottom(xLeftFn).ticks(x.axisTicks);

    const xRightAxis =
      // @ts-ignore
      x.axis === 'top'
        ? // @ts-ignore
          axisTop(xRightFn).ticks(x.axisTicks)
        : // @ts-ignore
          axisBottom(xRightFn).ticks(x.axisTicks);

    const xRightAxisG = g.append('g').attr('class', 'right-axis--x axis ');

    g.append('line')
      .attr('x1', xLeftFn(0))
      .attr('x2', xLeftFn(0))
      .attr('y1', margin.top)
      .attr('y2', height - margin.bottom)
      .attr('class', 'stroke-current stroke-1');

    const yFn = scaleBand()
      // @ts-ignore
      .domain(data.map((d) => d[y.key]))
      .range([margin.top + padding.top, height - padding.bottom - margin.bottom])
      .padding(paddingBar);

    leftSeries.map((column, i) => {
      const barsG = g.append('g');
      const columns = leftSeries.filter((_, idx) => idx >= i).map((c) => c.key);
      // const bars =
      barsG
        .selectAll('g')
        .data(data)
        .enter()
        .append('rect')
        .attr('class', `${column.className} fill-current`)
        // @ts-ignore
        .attr('y', (d) => yFn(d[y.key]))
        .attr('x', xLeftFn(0))
        .attr('width', 0)
        .attr('height', yFn.bandwidth())
        .transition()
        .duration(1000)
        // @ts-ignore
        .attr('x', (d) => xLeftFn(sum(columns.map((c) => d[c]))))
        .attr(
          'width',
          // @ts-ignore
          (d) => xLeftFn(0) - xLeftFn(sum(columns.map((c) => d[c])))
        );
    });
    rightSeries.map((column, i) => {
      const barsG = g.append('g');
      const columns = rightSeries
        .filter((_, idx) => idx >= i)
        .map((c) => c.key);
      // const bars =
      barsG
        .selectAll('g')
        .data(data)
        .enter()
        .append('rect')
        .attr('class', `${column.className} fill-current`)
        .attr('x', (d) => xRightFn(0))
        .attr('z-index', 100 - i)
        // @ts-ignore
        .attr('y', (d) => yFn(d[y.key]))
        .attr('height', yFn.bandwidth())
        .transition()
        .duration(1000)
        .attr(
          'width',
          // @ts-ignore
          (d) => xRightFn(sum(columns.map((c) => d[c]))) - xRightFn(0)
        );
    });

    // Draw axis

    xRightAxisG
      .attr(
        'transform',
        // @ts-ignore
        `translate(0, ${x.axis === 'top' ? margin.top : height - margin.bottom})`
      )
      .call(xRightAxis);

    const xLeftAxisG = g.append('g').attr('class', 'left-axis--x axis ');

    xLeftAxisG
      .attr(
        'transform',
        // @ts-ignore
        `translate(0, ${x.axis === 'top' ? margin.top : height - margin.bottom})`
      )
      .call(xLeftAxis);

    const yAxis = y.direction === 'left' ? axisRight(yFn) : axisLeft(yFn);
    const yAxisG = g
      .append('g')
      .attr('class', 'yAxis axis')
      .attr(
        'transform',
        `translate(${y.direction === 'left' ? width : margin.left},0)`
      );

    yAxisG.call(yAxis);
  },[data, x, padding, margin, paddingBar, y]);

  useEffect(() => {
    refreshChart();
    return () => {
      selectAll('.tooltip').remove();
    };
  }, [data]);

  /* eslint-enable */

  return (
    <svg
      id={id}
      className={mergeTailwindClasses(defaultChartClassNames, className)}
    />
  );
};

export default SpineChart;
