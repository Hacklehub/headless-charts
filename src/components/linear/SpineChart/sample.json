[
  {
    "name": "Product A",
    "value2": 6,
    "value3": 7,
    "value4": 8
  },
  {
    "name": "Product B",
    "value2": 4,
    "value3": 6,
    "value4": 8
  },
  {
    "name": "Product C",
    "value2": 1,
    "value3": 2,
    "value4": 3
  }
]
