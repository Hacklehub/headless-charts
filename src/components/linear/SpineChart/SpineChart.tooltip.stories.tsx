/* eslint-disable @typescript-eslint/no-explicit-any */
import { Meta, StoryObj } from '@storybook/react';

import SpineChart from '.';
import data from './sample.json';

const meta: Meta<typeof SpineChart> = {
  title: 'Linear/SpineChart/Tooltips',
  component: SpineChart,
  tags: ['autodocs'],
};

export default meta;

type Story = StoryObj<typeof SpineChart>;

export const Tooltip: Story = {
  args: {
    data,
    id: 'tooltip',
    y: {
      key: 'name',
      axis: 'left',
    },
    x: [
      { key: 'value2', direction: 'left' },
      { key: 'value3', direction: 'right' },
      { key: 'value4', direction: 'left' },
    ],
    tooltip: {
      keys: ['value2', 'value3', 'value4'],
      className: 'bg-gray-800 text-white p-2 rounded-md shadow-md',
    },
  },
};

export const TooltipCustomHtml: Story = {
  args: {
    ...Tooltip.args,
    id: 'tooltip-custom-html',
    tooltip: {
      html: (d: any) => `
        <div class="bg-gray-800 text-white p-2 rounded-md shadow-md">
          <div class="font-bold">${d.name}</div>
          <div>${d.value2}</div>
          <div>${d.value3}</div>
          <div>${d.value4}</div>
        </div>
      `,
    },
  },
};
