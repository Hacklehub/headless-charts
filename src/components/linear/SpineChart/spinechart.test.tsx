import { describe, expect, it } from 'vitest';
import { render, screen } from '@testing-library/react';

import SpineChart from '.';

const data = [
  {
    name: 'Product A',
    value2: 6,
    value3: 7,
    value4: 8,
  },
  {
    name: 'Product B',
    value2: 4,
    value3: 6,
    value4: 8,
  },
  {
    name: 'Product C',
    value2: 1,
    value3: 2,
    value4: 3,
  },
];

const x = [
  { key: 'value2', direction: 'left' },
  { key: 'value3', direction: 'right' },
  { key: 'value4', direction: 'left' },
];

describe('SpinrChart', () => {
  it('renders Spine chart component', () => {
    render(
      <SpineChart
        data={data}
        id='spineChart'
        className=''
        x={x}
        y={{
          key: '',
          axis: 'middle',
          className: '',
        }}
      />
    );
    const spineChart = screen.getByTestId('spineChart');
    expect(spineChart).toMatchSnapshot();
  });

  it('renders Spine chart component', () => {
    render(
      <SpineChart
        data={data}
        id='spineChart'
        className=''
        x={x}
        y={{
          key: '',
          axis: 'middle',
          className: '',
        }}
      />
    );
    const spineChart = screen.getByTestId('spineChart');
    expect(spineChart).toBeInTheDocument();
  });
});
