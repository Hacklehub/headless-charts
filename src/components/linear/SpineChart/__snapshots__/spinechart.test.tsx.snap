// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SpinrChart > renders Spine chart component 1`] = `
<svg
  class="chart dark:bg-gray-800 dark:text-gray-50 h-64 lg:w-4/12 md:w-6/12 text-gray-900 w-full"
  data-testid="spineChart"
  id="spineChart"
>
  <g>
    <g
      class="right-axis--x axis"
      fill="none"
      font-family="sans-serif"
      font-size="10"
      text-anchor="middle"
      transform="translate(0, -40)"
    >
      <path
        class="domain"
        d="M40.5,6V0.5H-19.5V6"
        stroke="currentColor"
      />
      <g
        class="tick"
        opacity="1"
        transform="translate(40.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          0
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(31.928571428571434,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          2
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(23.35714285714286,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          4
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(14.785714285714283,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          6
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(6.214285714285719,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          8
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(-2.3571428571428594,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          10
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(-10.928571428571427,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          12
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(-19.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          14
        </text>
      </g>
      <text
        class="fill-current"
        text-anchor="end"
        x="0"
        y="30"
      >
        value3
      </text>
    </g>
    <line
      class="stroke-current stroke-1"
      x1="-20"
      x2="-20"
      y1="40"
      y2="-40"
    />
    <g>
      <rect
        class="undefined fill-current"
        height="43.07692307692308"
        width="0"
        x="-20"
        y="-21.53846153846154"
      />
      <rect
        class="undefined fill-current"
        height="43.07692307692308"
        width="0"
        x="-20"
        y="-21.53846153846154"
      />
      <rect
        class="undefined fill-current"
        height="43.07692307692308"
        width="0"
        x="-20"
        y="-21.53846153846154"
      />
    </g>
    <g>
      <rect
        class="undefined fill-current"
        height="43.07692307692308"
        width="0"
        x="-20"
        y="-21.53846153846154"
      />
      <rect
        class="undefined fill-current"
        height="43.07692307692308"
        width="0"
        x="-20"
        y="-21.53846153846154"
      />
      <rect
        class="undefined fill-current"
        height="43.07692307692308"
        width="0"
        x="-20"
        y="-21.53846153846154"
      />
    </g>
    <g>
      <rect
        class="undefined fill-current"
        height="43.07692307692308"
        x="40"
        y="-21.53846153846154"
        z-index="100"
      />
      <rect
        class="undefined fill-current"
        height="43.07692307692308"
        x="40"
        y="-21.53846153846154"
        z-index="100"
      />
      <rect
        class="undefined fill-current"
        height="43.07692307692308"
        x="40"
        y="-21.53846153846154"
        z-index="100"
      />
    </g>
    <g
      class="left-axis--x axis "
      fill="none"
      font-family="sans-serif"
      font-size="10"
      text-anchor="middle"
      transform="translate(0, -40)"
    >
      <path
        class="domain"
        d="M-19.5,6V0.5H40.5V6"
        stroke="currentColor"
      />
      <g
        class="tick"
        opacity="1"
        transform="translate(-19.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          0
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(-10.928571428571432,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          2
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(-2.3571428571428594,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          4
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(6.214285714285715,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          6
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(14.785714285714281,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          8
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(23.35714285714286,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          10
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(31.928571428571427,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          12
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(40.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          14
        </text>
      </g>
      <text
        class="fill-current"
        text-anchor="start"
        x="0"
        y="30"
      >
        value2, value4
      </text>
    </g>
    <g
      class="yAxis axis"
      fill="none"
      font-family="sans-serif"
      font-size="10"
      text-anchor="end"
      transform="translate(40,0)"
    >
      <path
        class="domain"
        d="M-6,40.5H0.5V-39.5H-6"
        stroke="currentColor"
      />
      <g
        class="tick"
        opacity="1"
        transform="translate(0,0)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        />
      </g>
      <text
        class="fill-current"
        x="-20"
        y="-40"
      />
    </g>
  </g>
</svg>
`;
